--- ../src-base/minecraft/net/minecraft/village/Village.java
+++ ../src-work/minecraft/net/minecraft/village/Village.java
@@ -41,7 +41,7 @@
     private final List<Village.VillageAggressor> villageAgressors = Lists.<Village.VillageAggressor>newArrayList();
     private int numIronGolems;
 
-    public Village()
+    private Village() // Paper - Nothing should call this - world needs to be set.
     {
         this.capabilities = net.minecraftforge.event.ForgeEventFactory.gatherCapabilities(this);
     }
@@ -83,7 +83,7 @@
             {
                 EntityIronGolem entityirongolem = new EntityIronGolem(this.world);
                 entityirongolem.setPosition(vec3d.x, vec3d.y, vec3d.z);
-                this.world.spawnEntity(entityirongolem);
+                this.world.spawnEntity(entityirongolem, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.VILLAGE_DEFENSE);
                 ++this.numIronGolems;
             }
         }
@@ -382,7 +382,12 @@
 
     private boolean isWoodDoor(BlockPos pos)
     {
-        IBlockState iblockstate = this.world.getBlockState(pos);
+        // Paper start
+        IBlockState iblockstate = this.world.getTypeIfLoaded(pos);
+        if (iblockstate == null) {
+            return false;
+        }
+        // Paper end
         Block block = iblockstate.getBlock();
 
         if (block instanceof BlockDoor)
