--- ../src-base/minecraft/net/minecraft/block/BlockFrostedIce.java
+++ ../src-work/minecraft/net/minecraft/block/BlockFrostedIce.java
@@ -32,13 +32,16 @@
 
     public void updateTick(World worldIn, BlockPos pos, IBlockState state, Random rand)
     {
+        if (!worldIn.paperConfig.frostedIceEnabled) return; // Paper - add ability to disable frosted ice
         if ((rand.nextInt(3) == 0 || this.countNeighbors(worldIn, pos) < 4) && worldIn.getLightFromNeighbors(pos) > 11 - ((Integer)state.getValue(AGE)).intValue() - state.getLightOpacity())
         {
             this.slightlyMelt(worldIn, pos, state, rand, true);
         }
         else
         {
-            worldIn.scheduleUpdate(pos, this, MathHelper.getInt(rand, 20, 40));
+            // Paper start - use configurable min/max delay
+            worldIn.scheduleUpdate(pos, this, MathHelper.getInt(rand, worldIn.paperConfig.frostedIceDelayMin, worldIn.paperConfig.frostedIceDelayMax));
+            // Paper end
         }
     }
 
